pkg/dns/gcp/gcp.go

```diff
@@ -0,0 +1,71 @@
+package gcp
+
+import (
+	"context"
+
+	iov1 "github.com/openshift/cluster-ingress-operator/pkg/api/v1"
+	"github.com/openshift/cluster-ingress-operator/pkg/dns"
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+
+	configv1 "github.com/openshift/api/config/v1"
+
+	gdnsv1 "google.golang.org/api/dns/v1"
+	"google.golang.org/api/option"
+)
+
+var (
+	_   dns.Provider = &Provider{}
+	log              = logf.Logger.WithName("dns")
+)
+
+type Provider struct {
+	// config is required input.
+	config Config
+	// dnsService provides DNS API access.
+	dnsService *gdnsv1.Service
+}
+
+type Config struct {
+	Project   string
+	UserAgent string
+}
+
+func New(config Config) (*Provider, error) {
+	dnsService, err := gdnsv1.NewService(context.TODO(), option.WithUserAgent(config.UserAgent))
+	if err != nil {
+		return nil, err
+	}
+
+	provider := &Provider{
+		config:     config,
+		dnsService: dnsService,
+	}
+
+	return provider, nil
+}
+
+func (p *Provider) Ensure(record *iov1.DNSRecord, zone configv1.DNSZone) error {
+	rrs := &gdnsv1.ResourceRecordSet{
+		Kind:    "dns#resourceRecordSet",
+		Name:    record.Spec.DNSName,
+		Rrdatas: record.Spec.Targets,
+		Type:    record.Spec.RecordType,
+	}
+	change := &gdnsv1.Change{Additions: []*gdnsv1.ResourceRecordSet{rrs}}
+	call := p.dnsService.Changes.Create(p.config.Project, zone.ID, change)
+	_, err := call.Do()
+	return err
+}
+
+func (p *Provider) Delete(record *iov1.DNSRecord, zone configv1.DNSZone) error {
+	rrs := &gdnsv1.ResourceRecordSet{
+		Kind:    "dns#resourceRecordSet",
+		Name:    record.Spec.DNSName,
+		Rrdatas: record.Spec.Targets,
+		Type:    record.Spec.RecordType,
+	}
+	change := &gdnsv1.Change{Deletions: []*gdnsv1.ResourceRecordSet{rrs}}
+	call := p.dnsService.Changes.Create(p.config.Project, zone.ID, change)
```

Comment:
Are deletions idempotent, or will deleting a non-existent record return an error?

Reply:
Deletion is also idempotent